File: residential-proxies/PHP/src/FileManager.php
<?php

declare(strict_types=1);

namespace Oxylabs\ResidentialApi;

class FileManager
{
    public function __construct(
        private string $inputDirectory,
        private string $outputDirectory,
        private ConsoleWriter $consoleWriter
    ) {
    }

    public function readUrlList(): array
    {
        $this->consoleWriter->writeln('Reading from the list...');

        $urlList = @file(sprintf('%s/%s', $this->inputDirectory, URL_LIST_NAME));
        if (!$urlList) {
            $this->consoleWriter->writelnError('Failed to read input file');
            exit(1);
        }

        return array_filter($urlList);
    }

    public function writeError(string $contents): void
    {
        $this->consoleWriter->writeln('ERROR: ' . $contents);

        $fileName = sprintf('%s/failed_requests.txt', $this->outputDirectory);
        file_put_contents($fileName, $contents . PHP_EOL, FILE_APPEND);
    }

    public function writeSuccess($position, string $contents): void
    {
        $fileName = sprintf( '%s/result_%d.html', $this->outputDirectory, $position);
        file_put_contents($fileName, $contents);
    }
}

File: shared-datacenter-proxies/CSharp/Oxylabs/FileManager.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.RegularExpressions;

namespace Oxylabs
{
    class FileManager
    {
        public static String FAILED_REQUESTS_TXT = "failed_requests.txt";
        public static String REGEX_PATTERN = @"^dc\.(?<country>\w{2})-?pr\.oxylabs\.io:\d+$";


        private ConsoleWriter consoleWriter;

        public FileManager(ConsoleWriter consoleWriter)
        {
            this.consoleWriter = consoleWriter;

            File.Create(FAILED_REQUESTS_TXT).Dispose();
        }

        public string[] ReadUrlList()
        {
            return this.ReadLinesFromFile(Settings.URL_LIST_NAME);
        }

        public Dictionary<string, string> ReadProxyMap()
        {
            var proxyList = this.ReadLinesFromFile(Settings.PROXY_LIST_NAME);
            if(proxyList == null)
            {
                throw new Exception("Could not read proxy list from file");
            }

            var proxyMap = new Dictionary<string, string>();

            foreach (var proxyUrl in proxyList)
            {
                var match = Regex.Match(proxyUrl, REGEX_PATTERN);
                if (match.Groups.Count == 2)
                {
                    var country = match.Groups[1].ToString().ToUpper();
                    proxyMap.Add(country, proxyUrl);
                    continue;
                }

                proxyMap.Add(Settings.DEFAULT_PROXY_INDEX_NAME, proxyUrl);
            }

            return proxyMap;

        }

        public string[] ReadLinesFromFile(string fileName)
        {
            this.consoleWriter.Writeln("Reading from the list...");

            try
            {
                return System.IO.File.ReadAllLines(fileName);
            }
            catch (Exception e)
            {
                this.consoleWriter.WritelnAndExit("Failed to read from file: " + e.Message);
                return null;
            }
        }

        public void WriteError(String contents)
        {

            try
            {
                using (var stream = File.Open(FAILED_REQUESTS_TXT, FileMode.Append, FileAccess.Write, FileShare.ReadWrite))
                {
                    using (var fs = new StreamWriter(stream))
                    {
                        fs.WriteLine(contents);
                    }
                }
            }
            catch (Exception e)
            {
                this.consoleWriter.WritelnError("Failed to write to file: " + e.Message);
            }
        }

        public void writeSuccess(int position, String contents)
        {
            try
            {
                using (var stream = File.Open("result_" + position + ".html", FileMode.Create, FileAccess.Write, FileShare.ReadWrite))
                {
                    using (var fs = new StreamWriter(stream))
                    {
                        fs.WriteLine(contents);
                    }
                }
            }
            catch (Exception e)
            {
                this.consoleWriter.WritelnError("Failed to write to file: " + e.Message);
            }
        }
    }
}
